File: androidApp/src/main/java/com/paulomoura/pokedexkmm/android/presentation/pokemondetail/PokemonDetailViewModel.kt
package com.paulomoura.pokedexkmm.android.presentation.pokemondetail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.paulomoura.pokedexkmm.common.Response
import com.paulomoura.pokedexkmm.common.constants.ErrorConstants
import com.paulomoura.pokedexkmm.domain.usecase.GetPokemonUseCase
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class PokemonDetailViewModel(private val getPokemonUseCase: GetPokemonUseCase, private val pokemonNumber: Int) : ViewModel() {

    private val _uiState = MutableStateFlow(PokemonDetailState())
    val uiState = _uiState.stateIn(viewModelScope, SharingStarted.WhileSubscribed(), PokemonDetailState())

    init {
        getPokemon()
    }

    private fun getPokemon() {
        viewModelScope.launch {
            getPokemonUseCase.getPokemon(pokemonNumber).onEach { response ->
                _uiState.value = when (response) {
                    is Response.Loading -> PokemonDetailState(loading = true)
                    is Response.Success -> response.data?.let { PokemonDetailState(pokemon = it) }
                        ?: PokemonDetailState(error = ErrorConstants.ERROR_POKEMON_DETAIL_STATE)
                    is Response.Error -> PokemonDetailState(error = ErrorConstants.ERROR_POKEMON_DETAIL_STATE)
                }
            }.collect()
        }
    }
}

File: iosApp/iosApp/pokemondetail/PokemonDetailViewModel.swift
import shared
import Combine
import KMPNativeCoroutinesCombine

class PokemonDetailViewModel : ObservableObject {
    
    @Published var uiState = PokemonDetailState.loading(false)
    private var cancellable: AnyCancellable? = nil
    
    init(pokemonNumber: Int32) {
        getPokemon(pokemonNumber: pokemonNumber)
    }
    
    private func getPokemon(pokemonNumber: Int32) {
        let useCase = UseCasesHelper().getPokemonUseCase
        let publisher = createPublisher(for: useCase.getPokemon(number: pokemonNumber))
        cancellable = publisher.receive(on: DispatchQueue.main).sink { completion in
            print("Received completion: \(completion)")
        } receiveValue: { value in
            switch(value) {
            case is ResponseLoading<Pokemon>:
                self.uiState = PokemonDetailState.loading(true)
            case is ResponseSuccess<Pokemon>:
                if let pokemon = value.data {
                    self.uiState = PokemonDetailState.pokemon(pokemon)
                } else {
                    self.uiState = PokemonDetailState.error(ErrorConstants().ERROR_POKEMON_DETAIL_STATE)
                }
            case is ResponseError<Pokemon>:
                self.uiState = PokemonDetailState.error(ErrorConstants().ERROR_POKEMON_DETAIL_STATE)
            default:
                break
            }
        }
    }
}
